### PR TITLE
GBFS 2.0

### PR DESCRIPTION
**Pull request by ubahnverleih**
* fixing missing fields GBFS 1.0 and 1.1 required anyway
* added non-static bike ID for use in GBFS 2.0
* genrate non-static bike ID in migrations for all bikes
* regenarate non-static bike ID on rent finish

### CODE REVIEW

File: bikesharing/migrations/0027_bike_non_static_bike_uuid.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+    dependencies = [
+        ('bikesharing', '0026_auto_20200605_1839'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='bike',
+            name='non_static_bike_uuid',
+            field=models.UUIDField(default=uuid.uuid4, help_text="A temporary ID used in public APIs, rotating it's value after each rent to protect users privacy.", null=True),
```
Comment (robbi5) [inline bikesharing/migrations/0027_bike_non_static_bike_uuid.py:17]: can we add an unique-constraint for this field?

Comment (robbi5) [reply bikesharing/migrations/0027_bike_non_static_bike_uuid.py:17]: ah, migration 0029

File: gbfs/views.py [hunk 2]
```diff
@@ -46,7 +48,18 @@ def gbfs(request):
 
 def gbfsSystemInformation(request):
     if request.method == "GET":
-        data = {"system_id": "TBD", "language": "TBD", "name": "TBD", "timezone": "TBD"}
+        bsp = preferences.BikeSharePreferences
+        lang = translation.get_language()
+        twoletterlangcode = lang.split("-")[0]
+        #print(dir(info.values))
```
Comment (robbi5) [inline gbfs/views.py:54]: remove debugging print

File: api/views.py [hunk 3]
```diff
@@ -228,6 +230,8 @@ def finish_rent(request):
 
     # set Bike status back to available
     rent.bike.availability_status = "AV"
+    # set new non static bike ID, so for GBFS observers can not track this bike
+    rent.bike.non_static_bike_uuid = uuid.uuid4()
     rent.bike.save()
```
Comment (robbi5) [inline api/views.py:235]: what happens if unique constraint fails?

File: gbfs/views.py [hunk 4]
```diff
@@ -46,7 +48,18 @@ def gbfs(request):
...
+        data = {
+            "system_id": bsp.gbfs_system_id,
+            "license_url": "https://creativecommons.org/publicdomain/zero/1.0/",
+            "language": twoletterlangcode,
```
Comment (robbi5) [inline gbfs/views.py:58]: language code in gbfs.json (line 25) has to match: "The language that will be used throughout the rest of the files. It must match the value in the system_information.json file."

File: gbfs/views.py [hunk 5]
```diff
@@ -46,7 +48,18 @@ def gbfs(request):
 
 def gbfsSystemInformation(request):
     if request.method == "GET":
-        data = {"system_id": "TBD", "language": "TBD", "name": "TBD", "timezone": "TBD"}
+        bsp = preferences.BikeSharePreferences
+        lang = translation.get_language()
+        twoletterlangcode = lang.split("-")[0]
```
Comment (robbi5) [inline gbfs/views.py:53]: personally, I wouldn't split the langcode. the gbfs spec uses IETF BCP 47 there, `language-region/variant` is defined there. it only has to be consistent in all our usage.

File: bikesharing/migrations/0027_bike_non_static_bike_uuid.py (hunk 1 above)
Comment (ubahnverleih) [reply bikesharing/migrations/0027_bike_non_static_bike_uuid.py:17]: yep, this is the recommended hack for prefilled fields. create, fill, make unique

File: api/views.py (hunk 3 above)
Comment (ubahnverleih) [reply api/views.py:235]: if i correctly undestand, if our randomness is OK, then chance of collision is 2^64. So the system might explode every 18,446,744,073,709,551,615 rents

File: gbfs/views.py (hunk 2 above)
Comment (ubahnverleih) [reply gbfs/views.py]: fixed in abfcd08

File: gbfs/views.py (hunk 5 above)
Comment (ubahnverleih) [reply gbfs/views.py]: fixed in  abfcd08

File: api/views.py (hunk 3 above)
Comment (ubahnverleih) [reply api/views.py:233]: now finish rent doesn't fail anymore, just no ratation of uuid in 80558d6

**Comment by robbi5** [comment]
Just read https://medium.com/@mobilitydata/whats-new-in-gbfs-v2-0-63eb46e6bdc4#2bc4 again. Maybe we should keep GBFS1.x (and make our api fully compatible) and provide 2.x at a seperate url.

**Comment by robbi5** [comment]
Current discussion: Our GBFS didn't validate with GBFS1 anyways. If our current usage (leaflet-gbfs _and_ opentripplanner) is okay with gbfs2, we can skip gbfs1 for now.

**Pull request merged**